### PR TITLE
Build archives using the K_DARWIN format when targeting osx

### LINKED ISSUE #1981: creating static libraries targeting macosx results in unusable .a files
```zig
export fn foo() void {}
```

```
zig build-lib foo.zig
```

```c
void foo(void);
int main() {foo(); return 0;}
```

```
cc -o test test.c foo.a
ld: warning: ignoring file foo.a, file was built for archive which is not the architecture being linked (x86_64): foo.a
```

The error message is bogus; the file is of course the correct architecture.

Suggestion here is to use `libtool -static -o libio.a io.o` rather than `ar`.:
https://stackoverflow.com/questions/30948807/static-library-link-issue-with-mac-os-x-symbols-not-found-for-architecture-x8

We will have to figure out how to do that using LLVM API.

A workaround for this bug is to use `build-obj` instead of `build-lib`. Indeed that's the workaround here:

https://github.com/ziglang/zig/blob/74a335c4ccd9eb4cfcc5cc6c01b633584c6bb6ba/src/link.cpp#L35-L42

Related issue regarding a new linker: #1535 
Related issue regarding dynamic libraries: #1982

### LINKED ISSUE #2421: Failing test suite with 0.4.0 release on MacOS
WIth Zig 0.4.0 built from source on macOS 10.14.4 (Mojave), I get the following error when running `bin/zig build --build-file ../build.zig test -Dskip-release`:
```
Assertion failed: ((uint64_t)loc % alignof(T) == 0 && "invalid pointer alignment"), function read64, file /Users/davor/projects/ziggly/zig/deps/lld/lib/ReaderWriter/MachO/MachONormalizedFileBinaryUtils.h, line 86.
The following command terminated unexpectedly:
/Users/davor/projects/ziggly/zig/build/bin/zig test /Users/davor/projects/ziggly/zig/test/standalone/static_c_lib/foo.zig --object /Users/davor/projects/ziggly/zig/test/standalone/static_c_lib/zig-cache/o/SUe_YiPh1_Ew6fiegQeR8ME86r6PNe-adxCtuWf8bVaQx3w_6ftZmy4uax5-vaqd/libfoo.a --cache-dir /Users/davor/projects/ziggly/zig/test/standalone/static_c_lib/zig-cache --name test -isystem /Users/davor/projects/ziggly/zig/test/standalone/static_c_lib/zig-cache/o/SUe_YiPh1_Ew6fiegQeR8ME86r6PNe-adxCtuWf8bVaQx3w_6ftZmy4uax5-vaqd -isystem /Users/davor/projects/ziggly/zig/test/standalone/static_c_lib
```
Is this a known failure, or a possible problem with my setup? There are 5,718 passing tests (counted by their printing "...OK").

### PR DESCRIPTION
**Pull request by LemonBoy**
Closes #2421
Closes #1981
Removes all the horrible linking hacks for macos
Enables stack probes on macos

While lld Mach-O supports has its own share of problems this time it was not his fault heh

### What's next
- Is `system_linker_hack` still needed? I don't have a macos machine so please check this out
- `disable_stack_probing` is not needed anymore and it's described as a workaround, we may want to re-purpose it as a proper flag (`-fstack-check` and `-fno-stack-check` to be consistent with gcc & friends maybe?)

### CODE REVIEW

**Comment by emekoi** [comment]
if you're knowledgeable in this area, could you explain [this](https://github.com/ziglang/zig/blob/master/src/zig_llvm.cpp#L873)?

**Comment by LemonBoy** [comment]
> if you're knowledgeable in this area, could you explain this?

Uh, there's not much to explain. The [lib format](https://docs.microsoft.com/en-us/windows/desktop/debug/pe-format#archive-library-file-format) used by Windows is exactly the same as your bog-standard GNU-style `ar` archive.

**Comment by dm1try** [comment]
FYI, I've tested these changes locally. the problem described in #2421 is fixed indeed.
but, for #1981 I have this error:
Screen Shot 2019-05-22 at 19 05 21

**Comment by LemonBoy** [comment]
> but, for #1981 I have this error

That's also happens on linux, but that's a different bug ;)

**Comment by andrewrk** [comment]
This is an exciting PR, can't wait to merge this

File: src/zig_llvm.cpp [hunk 1]
```diff
@@ -877,6 +877,7 @@ bool ZigLLVMWriteArchive(const char *archive_name, const char **file_names, size
         case ZigLLVM_Linux:
             kind = object::Archive::K_GNU;
             break;
+        case ZigLLVM_MacOSX:
```
Comment (andrewrk) [inline src/zig_llvm.cpp]: That was the bug?? I forgot to put `ZigLLVM_MacOSX` in there? :man_facepalming:

**Comment by andrewrk** [comment]
> Is `system_linker_hack` still needed? I don't have a macos machine so please check this out

It looks like this is only enabled when building with gcc, so this can be tested with a CI build of a pull request.

> `disable_stack_probing` is not needed anymore and it's described as a workaround, we may want to re-purpose it as a proper flag (`-fstack-check` and `-fno-stack-check` to be consistent with gcc & friends maybe?)

2c0280ba085893984007706fb40c7b291f43074d

**Pull request merged**